ZFSin/spl/module/spl/spl-vnode.c

```diff
@@ -925,8 +926,14 @@ int   vnode_unlink(vnode_t *vp)
 
 void   vnode_setunlink(vnode_t *vp)
 {
+	if (vp && VTOZ(vp) && VTOZ(vp)->z_pflags & (ZFS_IMMUTABLE | ZFS_NOUNLINK)) {
```

Comment:
Generally, there should be no ZFS logic in SPL, so we might need to move this test up into ZFS, and only call setunlink when we decide to do so.

Reply:
Since we call it so often, maybe make a zfs_setunlink() that does the test before calling vnode_setunlink.

Reply:
I'll give it a try. ☺
I just had another thought: Wouldn't it be better to just call zfs_zaccess_delete() directly in order to determine whether we are authorized to delete something? 
This is what the function is intended for, isn't it?

Reply:
I honestly dont know, do we call it? not sure I even call it in osx version

Reply:
it gets called via fsdispatcher->IRP_MJ_CLEANUP->delete_entry->zfs_remove or ->zfs_rmdir in zfs_vnops.c like this:
```
if ((error = zfs_zaccess_delete(dzp, zp, cr))) {
	goto out;
}
```
this led me to the error initially. But I think, instead of picking single conditions out of zfs_zaccess_delete() and getting the same error again if any other condition is satisfied, it would be better to just use the function itself to determine whether we are going to run into access denied in IRP_MJ_CLEANUP...